picoquic/cubic.c

```diff
@@ -229,195 +241,164 @@ static void picoquic_cubic_notify(
     path_x->is_cc_data_updated = 1;
 
     if (cubic_state != NULL) {
-        switch (cubic_state->alg_state) {
-        case picoquic_cubic_alg_slow_start:
-            switch (notification) {
+        switch (notification) {
+            /* RTT measurements will happen before acknowledgement is signalled */
             case picoquic_congestion_notification_acknowledgement:
-                cubic_update_bandwidth(path_x);
-                if (path_x->last_time_acked_data_frame_sent > path_x->last_sender_limited_time) {
-                    picoquic_hystart_increase(path_x, &cubic_state->rtt_filter, ack_state->nb_bytes_acknowledged);
-                    /* if cnx->cwin exceeds SSTHRESH, exit and go to CA */
-                    if (path_x->cwin >= cubic_state->ssthresh) {
-                        cubic_state->W_reno = ((double)path_x->cwin) / 2.0;
-                        path_x->is_ssthresh_initialized = 1;
-                        picoquic_cubic_enter_avoidance(cubic_state, current_time);
-                    }
+                switch (cubic_state->alg_state) {
+                    case picoquic_cubic_alg_slow_start:
+                        /* Increase cwin based on bandwidth estimation. */
+                        path_x->cwin = picoquic_cc_bandwidth_estimation(path_x);
+
+                        if (path_x->last_time_acked_data_frame_sent > path_x->last_sender_limited_time) {
+                            //if (path_x->bytes_in_transit > path_x->cwin) {
+                                path_x->cwin += picoquic_cc_slow_start_increase_ex(path_x, ack_state->nb_bytes_acknowledged, 0);
+
+                                /* if cnx->cwin exceeds SSTHRESH, exit and go to CA */
+                                if (path_x->cwin >= cubic_state->ssthresh) {
+                                    cubic_state->W_reno = ((double)path_x->cwin) / 2.0;
+                                    path_x->is_ssthresh_initialized = 1;
+                                    cubic_enter_avoidance(cubic_state, current_time);
+                                }
+                            //}
+                        }
+                        break;
+                    /* TODO discuss
+                     * picoquic_cubic_alg_recovery is not entered anyway
+                     */
```

Comment:
Yes, that's a bug. But let's not try fix it in this PR. There are additional complexities. For example, we may want to enter recovery if the delay increased too much (DCubic), and in that case ignore delay measurements until we get acks of new packets. Maybe something similar for ECN. We could get scenarios in which we first enter recovery because of delay increase, and then notice a packet loss. In that case, if we used different betas for delay and losses, we will need to reduce the cwin some more (for the more important beta). Let's do that in a separate PR.

Reply:
Added issue #1823 to track possible recovery bugs.